stm/stm.py

```diff
@@ -0,0 +1,418 @@
+
+
+@xr.register_dataset_accessor("stm")
+class SpaceTimeMatrix:
+    """
+    Space-Time Matrix
+    """
+
+    def __init__(self, xarray_obj):
+        self._obj = xarray_obj
+
+    def add_metadata(self, metadata):
+        """
+        Assign metadata to the STM.
+
+        Parameters
+        ----------
+        metadata : str or dict
+            input metadata
+
+        Returns
+        -------
+        xarray.Dataset
+            STM with assigned attributes.
+        """
+        self._obj = self._obj.assign_attrs(metadata)
+        return self._obj
+
+    def subset(self, method: str, **kwargs):
+        """
+        Select a subset of the STM
+
+        Parameters
+        ----------
+        method : str
+            Method of subsetting. Choose from "threshold", "density" and "polygon".
+            - threshold: select all points with a threshold criterion, e.g.
+                data_xr.stm.subset(method="threshold", var="thres", threshold='>1')
+            - density: select one point in every [dx, dy] cell, e.g.
+                data_xr.stm.subset(method='density', dx=0.1, dy=0.1)
+            - polygon: select all points inside a given polygon, e.g.
+                data_xr.stm.subset(method='polygon', polygon=path_polygon_file)
+                or
+                import geopandas as gpd
+                polygon = gpd.read_file(path_polygon_file)
+                data_xr.stm.subset(method='polygon', polygon=polygon)
+
+        Returns
+        -------
+        xarray.Dataset
+            A subset of the original STM.
+        """
+
+        match method:  # Match statements available only from python 3.10 onwards
+            case "threshold":
+                _check_threshold_kwargs(
+                    **kwargs
+                )  # Check is all required kwargs are available
+                if kwargs["threshold"][0] == "<":
+                    str_parts = kwargs["threshold"].partition("<")
+                    check_mult_relops(
+                        str_parts[2]
+                    )  # Check to ensure multiple relational operators are not present
+                    data_xr_subset = self._obj.where(
+                        self._obj[kwargs["var"]] < float(str_parts[2]), drop=True
+                    )
+                elif kwargs["threshold"][0] == ">":
+                    str_parts = kwargs["threshold"].partition(">")
+                    check_mult_relops(str_parts[2])
+                    data_xr_subset = self._obj.where(
+                        self._obj[kwargs["var"]] > float(str_parts[2]), drop=True
+                    )
+                else:
+                    raise Exception(
+                        "Suitable relational operator not found! Please check input"
+                    )
+            case "density":
+                check_density_kwargs(**kwargs)  # Check for all require kwargs
+                gdf = gpd.GeoDataFrame(
```

Comment:
The implementation of the 'density' subset method is elegant. A bit more in-line documentation of the algorithm would help the user immediately understand the logic.

Reply:
Thx, I added some in-line docs